### PR TITLE
KV support

### LINKED ISSUE #27: Create astral interface for secrets
in addition to certificates, we should support more generic secrets (kv)

### PR DESCRIPTION
**Pull request by suprjinx**
Support for KV secret storage and retrieval

Fixes #27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,19 +18,29 @@ This Rails app is most easily run and developed in its devcontainer.
...
 ```
 curl -X POST http://localhost:3000/certificates \
 -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJqb2huLmRvZUBleGFtcGxlLmNvbSIsIm5hbWUiOiJKb2huIERvZSIsImlhdCI6MTUxNjIzOTAyMiwiZ3JvdXBzIjpbImdyb3VwMSIsImdyb3VwMiJdLCJhdWQiOiJhc3RyYWwifQ.tfRLXmE_eq-piP88_clwPWrYfMAQbCJAeZQI6OFxZSI" \
 -H "Content-type: application/json" \
 -d "{ \"cert_issue_request\": { \"common_name\": \"example.com\" } }"
 ```
-4) Run the tests from devcontainer terminal:
+4) POST and GET /secrets to save and fetch a secret:
+```
+curl -X POST http://localhost:3000/secrets \
```
Comment (dave-gantenbein) [inline README.md:30]: Not a blocker, but can we look into something that will generate docs for us from the code? Rails must provide something like that.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ class ApplicationController < ActionController::API
   before_action :store_request_id
   rescue_from StandardError, with: :handle_standard_error
   rescue_from AuthError, with: :handle_auth_error
-  rescue_from BadRequestError, with: :handle_bad_request_error
+  rescue_from BadRequestError, with: :handle_bad_request
```
Comment (dave-gantenbein) [inline app/controllers/application_controller.rb:6]: Thank you :)

File: app/lib/services/secrets_service.rb [hunk 3]
```diff
@@ -1,10 +1,22 @@
 module Services
-  class CertificateService
+  class SecretsService
     class << self
       def issue_cert(cert_issue_request)
         impl.issue_cert(cert_issue_request)
       end
 
+      def kv_read(path)
```
Comment (dave-gantenbein) [inline app/lib/services/secrets_service.rb:8]: SRP calling, do we want to conflate certs and kvs? Services::KeyValue.read(path) looks more railsish, no?

File: test/integration/secrets_test.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    post secrets_path
+    assert_response :unauthorized
+  end
+
+  test "#create with faulty token (encoded with different signing key)" do
+    post secrets_path, headers: { "Authorization" => "Bearer #{jwt_unauthorized}" }
+    assert_response :unauthorized
+  end
+
+  test "#create or update a secret" do
```
Comment (dave-gantenbein) [inline test/integration/secrets_test.rb:14]: Did you mean to assert something here?

Comment (suprjinx) [reply test/integration/secrets_test.rb:14]: well the "create_secret" call is doing the test, but i then need to reuse that for show/delete

Comment (suprjinx) [reply test/integration/secrets_test.rb:14]: i can move the assertions from there to here

File: README.md (hunk 1 above)
Comment (suprjinx) [reply README.md:30]: There's a story in the backlog to make Openapi / swagger doc -- but I often dislike having a lot of documentation mixed into the source code (as javadoc, rdoc, etc).

File: app/lib/services/secrets_service.rb (hunk 3 above)
Comment (suprjinx) [reply app/lib/services/secrets_service.rb:8]: The idea was this is a facade to the backend "secrets" provider (Vault or KMS) -- so if we have multiple Services classes, each will need to swap in the provider. Would the VaultService also need to break apart?

File: README.md (hunk 1 above)
Comment (dave-gantenbein) [reply README.md:30]: Can't it extract something from the controllers and allowed_params?

File: app/lib/clients/vault.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      end
+
+      def kv_delete(path)
+        client.logical.delete("#{kv_mount}/data/#{path}")
+      end
+
+      private
+
+      def client
+        # TODO create a new token for the session
```
Comment (dave-gantenbein) [inline app/lib/clients/vault.rb:27]: Are there tickets for the TODOs in this PR?

Comment (suprjinx) [reply app/lib/clients/vault.rb:27]: i think that's been around for a while but I moved the file. I'll delete TODO and make tix

Comment (suprjinx) [reply app/lib/clients/vault.rb]: https://github.com/G-Research/astral/issues/34

File: app/lib/services/secrets_service.rb (hunk 3 above)
Comment (dave-gantenbein) [reply app/lib/services/secrets_service.rb]: Well. maybe but it doesn't seem like it would be an intensive refactor, so we can defer that until later.

**Review by dave-gantenbein — APPROVED** [review]

**Pull request merged**